docs/reference.md

```diff
@@ -0,0 +1,523 @@
+  cava (signal (Vec Bit (a + b)))` : Synthesizable multiplier with growth
+- `greaterThanOrEqual : forall a b : nat, signal (Vec Bit a) * signal (Vec Bit
+  b) -> cava (signal Bit)` : Comparison of unsigned bit vectors
+
+### Circuit Combinators
+
+#### Tuple Manipulation
+
+These are especially useful with monad [composition](#compose) (`_ >=> _`).
+
+- `fork2 : forall {A}, signal A -> cava (signal A * signal A)` : replicate a
+  signal (fork a wire in the circuit)
+- `first : forall {A B C}, (A -> cava C) -> (A * B) -> cava (C * B)` : apply a
+  function to the first element of a pair
+- `second : forall {A B C}, (B -> cava C) -> (A * B) -> cava (A * C)` : apply a
+  function to the first element of a pair
+- `swap : forall {A B}, (A * B) -> cava (B * A)` : swap elements of a pair
+- `dropr : forall {A B}, (A * B) -> cava A` : drop right element of a pair
+- `dropl : forall {A B}, (A * B) -> cava B` : drop left element of a pair
+- `pair_left : forall {A B C}, (A * (B * C)) -> cava (A * B * C)` : reassociate
+  input tuple (the `*` operator is left-associative, so `(A * B * C)` means `((A *
+  B) * C))`)
+- `pair_right : forall {A B C}, (A * B * C) -> cava (A * (B * C))` : reassociate
+  input tuple (the `*` operator is left-associative, so `(A * B * C)` means `((A *
+  B) * C))`)
+
+#### 4-sided Tile Combinators
+
+- `below : forall {A B C D E F G}, (A * B -> cava (D * G)) -> (G * C -> cava
+  (E * F)) -> (A * (B * C)) -> cava ((D * E) * F)` : given two two-input
+  two-output circuits, feed one output of the first circuit into the second
+  circuit
+- `col : forall {A B C}, (A * B -> cava (signal C * A)) -> A -> forall {n :
+  nat}, Vector.t B n -> cava (signal (Vec C n) * A)` : chain a two-input
+  two-output circuit with itself
+
+The diagrams below illustrate the `below` and `col` combinators visually.
+
+![diagrams of the below and col tile combinators](below_and_col.svg){: width="800px"}
+
+#### Tree Combinators
+
+- `tree : forall {t}, (signal t * signal t -> cava (signal t)) -> forall {n},
+  signal (Vec t n) -> cava (signal t)` : create a balanced binary tree from a
+  two-input circuit
+
+### Lookup Tables
+
+- `lut1 : (bool -> bool) -> signal Bit -> cava (signal Bit)` : 1-bit lookup;
+  given a "table" function that returns a boolean given one boolean, emulate the
+  function as a circuit. This is essentially a lookup in a table containing two
+  elements.
+- `lut2 : (bool -> bool -> bool) -> signal Bit * signal Bit -> cava (signal
+  Bit)` : 2-bit lookup; a lookup in a table with four elements.
+- `lut3 : (bool -> bool -> bool -> bool) -> signal Bit * signal Bit * signal Bit
+  -> cava (signal Bit)` : 3-bit lookup
+- `lut4` : 4-bit lookup
+- `lut5` : 5-bit lookup
+- `lut6` : 6-bit lookup
+
+### Netlist-specific
+
+- `buf_gate : signal Bit -> cava (signal Bit)` TODO: describe
+- `localSignal : forall t, signal t -> cava (signal t)` TODO: describe
+- `instantiate` TODO: describe
```

Comment:
- `instantiate` similarly to `blackbox` will instantiate a module in SystemVerilog rather than inlining it. This is functionally equivalent to using the `instantiated` component directly, but produces a hierarchical design which may be useful for tooling that consumes the SystemVerilog.
- `buf_gate` generates a `buf` primitive in the SystemVerilog. It can have a configurable delay but currently we emit a 0 delay. In general explicit buffering can help with timing, but I'm not sure exactly how the `buf` SystemVerilog primitive gets interpreted (or if it is effectively a noop), I have previously only used explicit vendor buffering primitives.  
- `localSignal` makes sure that the current `Signal` has an explicit net numbering, rather than being an expression of `Signals`. This prevents an expression later being given multiple net numbers resulting in duplicated logic.

Reply:
`localSignal` binds an r-value to an l-value by giving a name to an expression. This is useful when this expression is used in several places to avoid circuits being duplicated.